### PR TITLE
fix duplicate logging entries

### PR DESCRIPTION
**Pull request by xoxys**
Background: https://github.com/Pylons/waitress/issues/151

For `waitress` it's required to add a log handler to the root logger (which is recommended in general) otherwise `waitress` will add its own handler. As a result, there exists a waitress handler and a flask handler which are both used for logging.

Result with debug log level:

```Shell
❯ python3 src/wopiserver.py
{"time": "2022-10-11T11:18:20.548", "host": "superion", "level": "INFO", "process": "wopiserver", "module": "wopiserver", "msg": "WOPI Server starting in unsecure/embedded mode", "port": "8880", "wopiurl": "https://your-wopi-server.org:8443", "version": "git"}
{"time": "2022-10-11T11:18:25.721", "host": "superion", "level": "DEBUG", "process": "wopiserver", "module": "wopiserver", "msg": "Accessed index page", "client": "127.0.0.1"}
DEBUG:wopiserver:"module": "wopiserver", "msg": "Accessed index page", "client": "127.0.0.1"
{"time": "2022-10-11T11:18:26.087", "host": "superion", "level": "DEBUG", "process": "wopiserver", "module": "wopiserver", "msg": "Accessed index page", "client": "127.0.0.1"}
DEBUG:wopiserver:"module": "wopiserver", "msg": "Accessed index page", "client": "127.0.0.1"
``` 

To fix it, this PR is removing the flask log handler and adding a proper handler to the root logger if waitress is used.

Result:

```Shell
❯ python3 src/wopiserver.py
{"time": "2022-10-11T11:20:30.725", "host": "superion", "level": "INFO", "process": "wopiserver", "module": "wopiserver", "msg": "WOPI Server starting in unsecure/embedded mode", "port": "8880", "wopiurl": "https://your-wopi-server.org:8443", "version": "git"}
{"time": "2022-10-11T11:20:38.280", "host": "superion", "level": "DEBUG", "process": "wopiserver", "module": "wopiserver", "msg": "Accessed index page", "client": "127.0.0.1"}
{"time": "2022-10-11T11:20:38.639", "host": "superion", "level": "DEBUG", "process": "wopiserver", "module": "wopiserver", "msg": "Accessed index page", "client": "127.0.0.1"}
{"time": "2022-10-11T11:20:39.029", "host": "superion", "level": "DEBUG", "process": "wopiserver", "module": "wopiserver", "msg": "Accessed index page", "client": "127.0.0.1"}
```

### CODE REVIEW

File: src/wopiserver.py [hunk 1]
```diff
@@ -108,6 +108,9 @@ def init(cls):
                 + hostname + '", "level": "%(levelname)s", "process": "%(name)s", %(message)s}',
                 datefmt='%Y-%m-%dT%H:%M:%S'))
             cls.app.logger.handlers = [loghandler]
+            if cls.config.get('general', 'internalserver', fallback='flask') == 'waitress':
+                cls.log.logger.handlers.clear()
+                logging.getLogger().handlers = [loghandler]
```
Comment (glpatcern) [inline src/wopiserver.py:113]: Just for polishing the code, why not dropping line 110 and do:
```
if <waitress>:
    logging.getLogger().handlers = [loghandler]
else:
    cls.app.logger.handlers = [loghandler]
```

**Review by glpatcern — CHANGES_REQUESTED** [review]
Looks fine, just a non-semantic change. As a matter of fact, we do see a few duplicate logs, only at the start of the process in `flask` mode. I wonder if something could be done (though it's very minor).

File: src/wopiserver.py (hunk 1 above)
Comment (xoxys) [reply src/wopiserver.py:113]: That doesn't work as flask has automatically registered a logger at `app.logger`.

**Comment by xoxys** [comment]
> As a matter of fact, we do see a few duplicate logs, only at the start of the process in flask mode. I wonder if something could be done (though it's very minor).

The first log entry `WOPI Server starting in unsecure/embedded mode` is duplicated because Flask immediately restart the server, but that's related to the debug mode behavior of Flask and IMO not a logging issue.

image

File: src/wopiserver.py (hunk 1 above)
Comment (glpatcern) [reply src/wopiserver.py:113]: I see, so the code should actually be:
```
if <waitress>:
    cls.log.logger.handlers.clear()
    logging.getLogger().handlers = [loghandler]
else:
    cls.app.logger.handlers = [loghandler]
```
My point is to make it straight out of the code that `cls.app.logger.handlers = [loghandler]` applies to Flask only.

**Comment by glpatcern** [comment]
> The first log entry WOPI Server starting in unsecure/embedded mode is duplicated because Flask immediately restart the server, but that's related to the debug mode behavior of Flask and IMO not a logging issue.

Right, thanks for that - and indeed not really an issue for this PR.

File: src/wopiserver.py (hunk 1 above)
Comment (xoxys) [reply src/wopiserver.py:112]: Ok, changed.

**Review by glpatcern — APPROVED** [review]

**Comment by xoxys** [comment]
@glpatcern Could we get a tag for this change?

**Comment by glpatcern** [comment]
@xoxys sure, I wanted to tag today after merging two more completed PRs.

**Comment by xoxys** [comment]
Cool, thanks!

**Pull request merged**